### PR TITLE
RD2: Display Elevate credit card widget

### PR DESCRIPTION
**Pull request by RadSFDeveloper**
Display Elevate credit card widget when a user creates a new Recurring Donation and select "Credit Card" as the payment method on an org connected to the Elevate payment services.

# Critical Changes

# Changes
The new `PS_ElevateService` and `UTIL_AuraEnabledCommon` Apex class should be added to required profiles' enabled classes.

# Issues Closed

# Community Ideas Delivered

# New Metadata
- Apex Class:
    - PS_ElevateService
    - UTIL_AuraEnabledCommon
<br/>

- Lightning Web Component:
    - psElevateTokenHandler
    - rd2ElevateCreditCardForm
<br/>

- Custom Label:
    - RD2_ElevateDisableButtonLabel
    - RD2_ElevateDisabledMessage
    - RD2_ElevateEnableButtonLabel

# Deleted Metadata

### CODE REVIEW

**Comment by RadSFDeveloper** [comment]
Tracking [W-8010251](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000008UhPQIA0/view)

File: src/classes/PS_ElevateService.cls
Comment (RadSFDeveloper) [inline src/classes/PS_ElevateService.cls:36]: There are common Payment Services methods that will be used by RD2 and probably other features, but currently they are contained in `GE` classes. I *copied* the common methods into a new Apex class. I have not extracted them from the `GE` classes, but I would suggest they are updated to use the common code in another WI. 

This new controller should be added to the required profiles' enabled Apex classes.

File: src/classes/PS_ElevateService.cls-meta.xml
Comment (RadSFDeveloper) [inline src/classes/PS_ElevateService.cls-meta.xml:3]: @force2b , should we be using API version 49 for new classes/LWCs?

File: src/classes/RD2_EntryFormController.cls
Comment (RadSFDeveloper) [inline src/classes/RD2_EntryFormController.cls:74]: The `PilotEnabled` feature parameter and its override on a sandbox will be used to enable Elevate credit card widget *in addition* to having the org connected to Elevate. When functionality becomes GA, the `PilotEnabled` check can be removed.

File: src/labels/CustomLabels.labels
Comment (RadSFDeveloper) [inline src/labels/CustomLabels.labels:1621]: @sam-knox , can you please verify three new labels applicable to RD2 entry form explicitly?

File: src/lwc/geFormRenderer/geFormRenderer.js
Comment (RadSFDeveloper) [inline src/lwc/geFormRenderer/geFormRenderer.js:1020]: @ptewson-sfdo , the GE form error handling was changed to account to an error response not containing message and index, so that one error returned from the iframe error response can be displayed.

File: src/lwc/psElevateTokenHandler/psElevateTokenHandler.js
Comment (RadSFDeveloper) [inline src/lwc/psElevateTokenHandler/psElevateTokenHandler.js:5]: The namespace is verified by checking `DataImport__c.Payment_Authorization_Token__c`. This is valid even when the user does not have read access to the NPSP Data Import SObject or read access to the field. @rponti-sforg , please verify  when testing the feature for a user that is not an admin.

File: src/lwc/psElevateTokenHandler/psElevateTokenHandler.js
Comment (RadSFDeveloper) [inline src/lwc/psElevateTokenHandler/psElevateTokenHandler.js:28]: @ptewson-sfdo , a new Elevate token handler was created for the common methods. I have not changed the GE form, but in the case the Visualforce origin handling is changed and this PR is not merged into the `master` yet, feel free to copy the class and use/update the common methods.

File: src/lwc/psElevateTokenHandler/psElevateTokenHandler.js
Comment (RadSFDeveloper) [inline src/lwc/psElevateTokenHandler/psElevateTokenHandler.js:137]: @ptewson-sfdo , the resolve contains the token, no field setting in the new common token handler class.

File: src/lwc/rd2ElevateCreditCardForm/rd2ElevateCreditCardForm.html
Comment (RadSFDeveloper) [inline src/lwc/rd2ElevateCreditCardForm/rd2ElevateCreditCardForm.html:21]: @rjanjanam , can you please verify the `data-qa-locators` make sense?

**Review by RadSFDeveloper — COMMENTED** [review]
@ptewson-sfdo , @force2b , @rponti-sforg , @andrewyu-salesforce 

Adding card holder input fields (as placeholders, they are not used by changes in this WI) are left, but please can you start the review/testing in meantime and I will make additional changes (if any) based on the code review. Please note this PR will be merged into `feature/RD2` and not into `master` this sprint. I also assume the `feature/RD2` will not be merged into `master` in the next sprint, but this will be confirmed during the next sprint planning. The feature is protected from being displayed to users even when the org is the Elevate customer by utilizing the `PilotEnabled` feature parameter.

**Review by sam-knox — APPROVED** [review]

**Comment by sam-knox** [comment]
@RadSFDeveloper - Per the notes in Changes, do all NPSP customers need to assign `PS_ElevateService` to profiles, or only customers who want to use Elevate?

**Comment by RadSFDeveloper** [comment]
> @RadSFDeveloper - Per the notes in Changes, do all NPSP customers need to assign `PS_ElevateService` to profiles, or only customers who want to use Elevate?

Great question @sam-knox , the service call will be made only when the org is connected to Elevate.

**Comment by RadSFDeveloper** [comment]
@rjanjanam , can you please verify [robot test failures](https://sfdo-metaci.herokuapp.com/builds/312242/tests) are not caused by this PR changes?  Thank you!

File: src/classes/PS_ElevateService.cls
Comment (force2b) [reply src/classes/PS_ElevateService.cls:36]: @RadSFDeveloper Because this class queries the `Organization` object, which is a System object, I think the class should be defined as either `without` or `inherited sharing` to avoid the same issue we ran into with the ApexAsyncJob system object. I will note that other classes that query this object are mostly implied inherited. There is one class that uses `with sharing`, so I suspect it would work even if left as is.

File: src/classes/PS_ElevateService.cls-meta.xml
Comment (force2b) [reply src/classes/PS_ElevateService.cls-meta.xml:3]: Yeah. I think we should

File: src/lwc/psElevateTokenHandler/psElevateTokenHandler.js-meta.xml
Comment (force2b) [inline src/lwc/psElevateTokenHandler/psElevateTokenHandler.js-meta.xml:4]: A description should be included on any new LWC's.

File: src/lwc/rd2ElevateCreditCardForm/rd2ElevateCreditCardForm.css
Comment (force2b) [inline src/lwc/rd2ElevateCreditCardForm/rd2ElevateCreditCardForm.css:19]: What if the User changes the theme in their org? Normally that would cause other colors to shift throughout the app, but here the background color is this section would remain the same even in an alternate theme. Is there a SLDS style that can be used to define this or a similar color so it's not hardcoded/fixed?

**Review by force2b — COMMENTED** [review]
Looking very nice. I need to do a second review of this and test in an org, but wanted to get you my initial comments asap.

File: src/classes/PS_ElevateService.cls
Comment (ptewson-sfdo) [reply src/classes/PS_ElevateService.cls:36]: Depending on timing, I wonder if we (Honeygems) might want to put in a PR as a child of this work to pivot to the new PS-prefixed variants. At the time we initially built this there wasn't enough intel to know if we were really going to share concepts like "isElevateCustomer" outside of Gift Entry, but since it is now clear that we will and are sharing these functions, I agree that this switch makes sense. But if we gather all that work into this PR we can solve for it now, rather than waiting until this work merges when we will inevitably be focused on other things.

**Review by ptewson-sfdo — APPROVED** [review]
This looks good from my side! Will want to coordinate on this side of things but verified that the Gift Entry form continues to work and that the error handling appears to be restored. Thanks!

File: src/classes/PS_ElevateService.cls
Comment (RadSFDeveloper) [reply src/classes/PS_ElevateService.cls:36]: @patrick-yan-sf , feel free to create branches off the `feature/RD2` to incorporate work by Honeygems related to Elevate Payment Services based off the common classes. The `feature/RD2` will be main branch for Elevate commitment creation work.

File: src/lwc/rd2ElevateCreditCardForm/rd2ElevateCreditCardForm.css
Comment (RadSFDeveloper) [reply src/lwc/rd2ElevateCreditCardForm/rd2ElevateCreditCardForm.css]: Excellent point! The change has been made to use [design tokens](https://developer.salesforce.com/docs/atlas.en-us.lightning.meta/lightning/tokens_standard_force_base.htm) and the fallback colors (in the case token rendering does not work).

**Comment by RadSFDeveloper** [comment]
@force2b , changes have been applied. The `PS_ElevateService` has been changed to use `without sharing`.

File: src/classes/PS_ElevateService.cls
Comment (force2b) [inline src/classes/PS_ElevateService.cls:43]: Is this method used anywhere?

**Review by andrewyu-salesforce — APPROVED** [review]
@RadSFDeveloper The PR looks great to me. I like how you already have the implementation layout for the future WIs such as sending token and error handling.

File: src/classes/PS_ElevateService.cls
Comment (RadSFDeveloper) [reply src/classes/PS_ElevateService.cls:43]: It is not used currently (it did at some part of the RD2 credit widget but decided to use `RD2_ElevateIntegrationService.isIntegrationEnabled()`), but most importantly it was created so Gift Entry code referencing common Elevate methods can start using the `PS_ElevateService`. Should I remove it for now, and leave this to Honeygems to add when they refactor the code to use the common methods?
CC @ptewson-sfdo

Comment (force2b) [reply src/classes/PS_ElevateService.cls:43]: Thanks for the clarification. There are other parts of the code that return back the elevate status as part of wrapper class structure. Wondering if that's the pattern to stick with even for future code.

I suppose, along the same lines, the `getOrgDomainInfo()` method is more of a utility method than something  specific to PaymentServices, though it is being used specifically by PS right now. Rather than include the AuraEnabled method in this class, which would then require permissions granted, could there be a `UTIL_LWC` (or `LWC_Utilities') class that has exposed utility methods, even if those methods delegate to other service classes? That could also help with permissions going forward.

Comment (RadSFDeveloper) [reply src/classes/PS_ElevateService.cls:43]: I understand for `getOrgDomainInfo()` that is a generic even though it is not used outside of the Elevate wrapper currently, but I assume we will have many Elevate related AuraEnabled methods so we do need a class specific to Elevate. I would avoid mixing Elevate AuraEnabled and generic methods in one class.

Comment (force2b) [reply src/classes/PS_ElevateService.cls:43]: That's exactly my concern here. `getOrgDomainInfo` is generic and is in this class, while this class is meant to be specific to PaymentServices related methods.

Comment (force2b) [reply src/classes/PS_ElevateService.cls:43]: That noted, if this class is going to expose AuraEnabled methods we can leave `getOrgDomainInfo` in here for now and eventually move out. The benefit of moving now though is that we can add the new utility class permission to our unpackaged metadata config for new RD2 users.

Comment (RadSFDeveloper) [reply src/classes/PS_ElevateService.cls:43]: The new `UTIL_AuraEnabledCommon` has been created a generic utility class for common methods regardless of the LC/LWC domain. Its name and/or `PS_ElevateService` can be changed before merging into `master`, and we will discuss this with Honeygems next week at the dev meeting. CC @ptewson-sfdo

**Review by force2b — APPROVED** [review]
Thanks for making those changes @RadSFDeveloper. I'm approving with one remaining question - is a new unit test class needed for the `UTIL_AuraEnabledCommon` class?

**Comment by RadSFDeveloper** [comment]
> Thanks for making those changes @RadSFDeveloper. I'm approving with one remaining question - is a new unit test class needed for the `UTIL_AuraEnabledCommon` class?

Valid question. Tests do not exist for `getOrgDomainInfo` currently. It is a matter of executing the method and verifying the values are not null. I will create the test class, but if we change the name of the common util class, the test class name needs to change as well.

**Comment by RadSFDeveloper** [comment]
@rponti-sforg , Elevate credit card widget is ready for testing.

**Pull request merged**
